crates/wasmparser/src/validator/operators.rs

```diff
@@ -2715,8 +2715,27 @@ where
         Ok(())
     }
     fn visit_ref_eq(&mut self) -> Self::Output {
-        self.pop_operand(Some(RefType::EQ.nullable().into()))?;
-        self.pop_operand(Some(RefType::EQ.nullable().into()))?;
+        match (self.pop_ref()?, self.pop_ref()?) {
+            (Some(a), Some(b)) => {
+                const EQREF: RefType = RefType::EQ.nullable();
+                let shared_eqref = EQREF.shared().expect("eq is abstract");
+                if a != EQREF && a != shared_eqref {
+                    bail!(
+                        self.offset,
+                        "type mismatch: expected `{}` or `{}`",
+                        EQREF,
+                        shared_eqref
+                    );
+                }
+                if a != b {
+                    bail!(
+                        self.offset,
+                        "type mismatch: expected `ref.eq` types to match"
+                    );
+                }
+            }
```

Comment:
Both of these checks are incorrect: the first should be a subtype check and I now am wondering as well if `a` and `b` can actually be different types.

Reply:
If it's common enough it's probably worth having a helper for, but as a one-off here what about:

```rust
if self.top_of_stack_is_shared_ref() {
    // two pops with EQ.shared()
} else {
    // two pops with EQ
}
```

Reply:
Looks like I'll also need it for `i31.get_s`, `i31.get_u`, `array.len`, `any.convert_extern`, and `extern.convert_any`. I've been staring at `pop_operand`, `_pop_operand`, and `pop_ref` but I don't see a nice way to incorporate this logic there. Should I just do what you suggested in all the spots?

Reply:
Perhaps a version of `pop_ref` which takes a `enum Shared { Yes, No, Maybe }` argument or similar? Given the use in other instructions a helper does seem in order

Reply:
Ok, take a look at b7d1e6f. There are certainly other ways to refactor things here so feel free to suggest alternate routes; I tried to match what was already there but it might be time for a more wholesale change.